compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt

```diff
@@ -0,0 +1,172 @@
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.FirSymbolOwner
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.declaration.isEnumEntryInitializer
+import org.jetbrains.kotlin.fir.analysis.checkers.getContainingClass
+import org.jetbrains.kotlin.fir.analysis.checkers.outerClass
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+
+object FirUninitializedEnumChecker : FirQualifiedAccessChecker() {
+    // Initialization order: member property initializers, enum entries, companion object (including members in it).
+    //
+    // When JVM loads a class, the corresponding class initializer, a.k.a. <clinit>, is executed first.
+    // Kotlin (and Java as well) converts enum entries as static final field, which is initialized in that <clinit>:
+    //   enum class E(...) {
+    //     E1, E2, ...
+    //   }
+    //     ~>
+    //   class E {
+    //     final static E1, E2, ...
+    //     static { // <clinit>
+    //       E1 = new E(...)
+    //       ...
+    //     }
+    //   }
+    //
+    // Note that, when initializing enum entries, now we call the enum class's constructor, a.k.a. <init>, to initialize non-final
+    // instance members. Therefore, if there is a member property in the enum class, and if that member has an access to enum entries,
+    // that is an illegal access since enum entries are not yet initialized:
+    //   enum class E(...) {
+    //     E1, E2, ...
+    //     val m1 = E1
+    //   }
+    //     ~>
+    //   class E {
+    //     E m1 ...
+    //     E(...) { // <init>
+    //       m1 = E1
+    //     }
+    //     final static E1, E2, ...
+    //     static { // <clinit>
+    //       E1 = new E(...)
+    //       ...
+    //     }
+    //   }
+    //
+    // A companion object is desugared to a static final singleton, and initialized in <clinit> too. However, enum lowering goes first,
+    // or in other words, companion object lowering goes last. Thus, any other things initialized in <clinit>, including enum entries,
+    // should not have access to companion object and members in it.
+    //
+    // See related discussions:
+    // https://youtrack.jetbrains.com/issue/KT-6054
+    // https://youtrack.jetbrains.com/issue/KT-11769
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
```

Comment:
Note that FE 1.0 performs this analysis as a part of CFA. It's a bit strange for me to see here separate logic about preceding enum entries and so. May be it's wrong architecturally.

@demiurg906 , what do you think about it?

Reply:
Actually I like that this checker is not part of CFA, because there is no such thing as graph of static initialization in FIR CFA, and this checker exactly about it. I think that it may be moved to CFA after we add true statics to language, but for now this is only one case of such errors, so it's better to have small accurate separate checker instead of adding hacks to CFA or building graph for static sections without proper design

Reply:
Ok